Motoko low_wasm_memory example (#1139)
dfinity/examples: rust/low_wasm_memory/README.md

# Low Wasm memory hook

The Internet Computer can automatically execute a special type of function called low Wasm memory hook, which runs when the available Wasm memory of the canister falls below the 'wasm_memory_threshold'.

> This Rust example demonstrates the ways of using low Wasm memory hook on the Internet Computer. If you're interested in how this example is implemented in Motoko, check out the [Motoko version of the Low Wasm memory hook example](../../motoko/low_wasm_memory).

Reviewer: ```suggestion
This Rust example demonstrates using the low Wasm memory hook on ICP. If you're interested in how this example is implemented in Motoko, check out the [Motoko variation](../../motoko/low_wasm_memory).
```

- [x] Install the [IC SDK](https://internetcomputer.org/docs/current/developer-docs/getting-started/install).
- [x] Clone the example dapp project: `git clone https://github.com/dfinity/examples`

> ## Example: Low Wasm memory hook

Reviewer: ```suggestion
## Using the example
```

In this example, the canister will periodically increase its memory usage (as defined in the `heartbeat` function) until the low Wasm memory hook is run
when the memory usage exceeds the `wasm_memory_threshold`.

> - ### Step 1: Setup project environment

Reviewer: ```suggestion
- #### Step 1: Setup project environment
```

> Navigate into the folder containing the project's files and start a local PocketIC instance with the command:

Reviewer: ```suggestion
Navigate into the folder containing the project's files and start a local developer environment with the commands:
```

This terminal will stay blocked, printing log messages, until the `Ctrl+C` is pressed or `dfx stop` command is run.

> - ### Step 2: Open another terminal window in the same directory:

Reviewer: ```suggestion
- #### Step 2: Open another terminal window in the same directory:
```

```sh
cd examples/rust/low_wasm_memory
```

> - ### Step 3: Create a new canister

Reviewer: ```suggestion
- #### Step 3: Create a new canister
```

```sh
dfx canister create low_wasm_memory_hook
```

Example output:

```sh
Created a wallet canister on the "local" network for user "default" with ID "uqqxf-5h777-77774-qaaaa-cai"
low_wasm_memory_hook canister created with canister id: uxrrr-q7777-77774-qaaaq-cai
```

> - ### Step 4: Update canister settings

Reviewer: ```suggestion
- #### Step 4: Update canister settings
```

Update canister settings:

```sh

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Author: Unfortunately this is by design, also in the test workflows. This is because motoko canisters use more memory at the start. This is to balance the wait time, with just 2MiB threshold we would need to wait too long for the memory to grow

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Reviewer: Ah okay, got it. Maybe a small note somewhere in the Motoko README about its increased memory usage would be helpful then?

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Author: Where do you think it would fit? I guess this might be confusing when someone compares both Motoko and Rust examples that the settings are different. But it might also be confusing to talk about the starting memory being larger for Motoko when someone just follows the Motoko example.

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Reviewer: Maybe under "Further learnings" adding something like "Motoko canisters utilize memory differently than canisters written in other languages. Learn more about Motoko memory management: (link to motoko docs)" ?

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Author: I'm not entirely how to explain the increased starting memory size for Motoko canisters apart from: "There is extra stuff in Motoko like garbage collection etc." We don't have anything relevant in the docs about memory management. I've found [canister-maintenance/memory](https://internetcomputer.org/docs/motoko/main/canister-maintenance/memory) which sounds promising by the link but it's just a doc for lowmemory hook 😄 I've added it to the Further learnings because it's anyway relevant to this example.
I've added an extra sentence to both readmes describing the starting memory after deployment which should give a reason for the specific values in the settings.
What do you think?

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Reviewer: LGTM! We can revisit the docs links once the new Motoko docs are live :)

> dfx canister update-settings low_wasm_memory_hook --wasm-memory-limit 5000000 --wasm-memory-threshold 3000000

Author: Ok, thank you for the reviews!

```

> - ### Step 5: Compile and deploy the `low_wasm_memory_hook` canister:

Reviewer: ```suggestion
- #### Step 5: Compile and deploy the `low_wasm_memory_hook` canister:
```

Example output:

```sh
Deploying: low_wasm_memory_hook
All canisters have already been created.
Building canister 'low_wasm_memory_hook'.
...
Installed code for canister low_wasm_memory_hook, with canister ID uxrrr-q7777-77774-qaaaq-cai
Deployed canisters.
URLs:
  Backend canister via Candid interface:
    low_wasm_memory_hook: http://127.0.0.1:4943/?canisterId=u6s2n-gx777-77774-qaaba-cai&id=uxrrr-q7777-77774-qaaaq-cai
```

After the deployment, the memory usage periodically increases as defined in the `heartbeat` function.

> With the `dfx canister update-settings` command we set the 'wasm_memory_limit' to 5MB and 'wasm_memory_threshold' to 3MB.

Reviewer: ```suggestion
The `dfx canister update-settings` command sets the 'wasm_memory_limit' to 5MiB and 'wasm_memory_threshold' to 2MiB.
```

Author: ```suggestion
The `dfx canister update-settings` command sets the 'wasm_memory_limit' to 5MiB and 'wasm_memory_threshold' to 3MiB.
```

> Hence whenever the Wasm memory used by the canister is above 2MB (in other words, the remaining Wasm memory is less than 'wasm_memory_threshold') the low Wasm memory hook will be triggered.

Author: ```suggestion
Hence whenever the Wasm memory used by the canister is above 2MiB (in other words, the remaining Wasm memory is less than 'wasm_memory_threshold'), the low Wasm memory hook will be triggered.
```

Example output:

```sh
Canister status call result for low_wasm_memory_hook.
Status: Running
Controllers: 3apfx-fn75o-xtwmf-svjzg-hu5xt-bg2dr-ubczq-uhvlq-2a5gf-ya4fn-dqe uqqxf-5h777-77774-qaaaa-cai
Memory allocation: 0 Bytes
Compute allocation: 0 %
Freezing threshold: 2_592_000 Seconds
Idle cycles burned per day: 14_902_151 Cycles
Memory Size: 1_458_248 Bytes
Balance: 2_996_988_117_866 Cycles
Reserved: 0 Cycles
Reserved cycles limit: 5_000_000_000_000 Cycles
Wasm memory limit: 5_000_000 Bytes

> Wasm memory threshold: 3_000_000 Bytes

Reviewer: ```suggestion
Wasm memory threshold: 2_000_000 Bytes
```

Module hash: 0x5f7571e87229a31fb1c3533479a2bdcef43cd0aaf3d33f852b88eab7ae72b3ae
Number of queries: 0
Instructions spent in queries: 0
Total query request payload size: 0 Bytes
Total query response payload size: 0 Bytes
Log visibility: controllers
```

> - ### Step 6: After a few seconds, observe the output of the `get_executed_functions_order` query:

Reviewer: ```suggestion
- #### Step 6: After a few seconds, observe the output of the `get_executed_functions_order` query:
```
